### ISSUE TITLE
RequestHook#onRequest sometimes not called

### ISSUE THREAD

**Issue by Klaster1** [DevExpress/testcafe#4616]
### What is your Test Scenario?
Create a test that has a `RequestHook` added in `test.before` or `fixture.beforeEach`. Open a page with `test.page` or `fixture.page`. 

I use this pattern to redirect/proxy XHR requests to instances of mock backend webserver , each on a separate port. The first request after page load is GET `/api/v1/user` that restores session, so when the issue happens it causes a test failure at the very start. The `-c 2` CLI option seems to increase the chance of this, especially in CI environment (Docker/TeamCity).
 
### What is the Current behavior?

When page sends an XHR, the hook does not trigger or call `onRequest` callback, this happens with varying frequency: always in reproducer, _sometimes_ in real test (which I can't disclose). This happens to both custom and default hooks, such as `RequestLogger`.

_If I use `t.navigateTo` instead of `test.page`, the hook works as expected._

After several debugging sessions, I managed to find out that the issue seems to happen because request hook event listeners are assigned after the request is made. These are the patches I used to monitor the request execution:

    // testcafe-hammerhead/lib/request-pipeline/index.js
    // line 59, sendDestinationRequest function
      if (ctx.reqOpts.url.includes('Einstein')) {
        console.log(
          Date.now(),
          'Session: ', ctx.session.id,
          'Has histeners: ', ctx.session.hasRequestEventListeners(),
          'Request to: ', ctx.reqOpts.url)
      }

    // testcafe-hammerhead/lib/session/index.js
    // line 215, addRequestEventListeners method
        console.log(Date.now(), 'Session: ', this.id, 'Adding listeners')
 
Example output:
```
 
> testcafe-issue-reproducer@1.0.0 test c:\dev\testcafe-issue-reproducer
> testcafe chrome:headless ./fixture.js

 Running tests in:
 - Chrome 80.0.3987.16 / Windows 10

 Reproducer
1577791293536 Session:  ADpQFeY1A Has histeners:  false Request to:  https://en.wikipedia.org/w/api.php?action=query&titles=Albert%20Einstein&prop=info&format=json
1577791293613 Session:  ADpQFeY1A Adding listeners
 × Fail

   1) AssertionError: expected 0 to deeply equal 1

      Browser: Chrome 80.0.3987.16 / Windows 10

         16 |    await t.addRequestHooks(t.ctx.logger = RequestLogger(/Einstein/))
         17 |})
         18 |('Fail', async t => {
         19 |    await t.expect(Selector('div').withText('Welcome').visible).ok()
         20 |    const {log} = await t.getBrowserConsoleMessages()
       > 21 |    await t.expect(t.ctx.logger.requests.length).eql(1)
         22 |})
         23 |
         24 |test
         25 |.clientScripts({content: SCRIPT, page: URL})
         26 |.before(async t => {

         at <anonymous> (c:\dev\testcafe-issue-reproducer\fixture.js:21:50)


1577791294629 Session:  ZeNvA0EMZ Adding listeners
1577791294925 Session:  ZeNvA0EMZ Has histeners:  true Request to:  https://en.wikipedia.org/w/api.php?action=query&titles=Albert%20Einstein&prop=info&format=json
 √ Pass


 1/2 failed (2s)
```

### What is the Expected behavior?
I expect request hook to work reliably independent of how page URL was specified. Request hooks should not miss any requests.
 
### What is your web application and your TestCafe test code?
[Minimal reproducer repo.](https://github.com/Klaster1/testcafe-issue-reproducer-1)

### Steps to Reproduce:

1. Clone the reproducer repo.
2. npm i
3. npm run test
 
### Your Environment details:
 
* testcafe version:                   1.7.1
* node.js version:                    v13.5.0
* command-line arguments:  testcafe chrome:headless ./fixture.js
* browser name and version: Chrome 80
* platform and version:          Windows 10 1909

**Comment by Dmitry-Ostashev** [comment]
The `test.before` and `fixture.beforeEach` hooks are called after the page is loaded. So, in your example, RequestHook is attached too late, after the script execution. I suggest you attach the request hook using the `test.requestHooks` method:

```
const logger = RequestLogger(/Einstein/);

fixture('Reproducer');

test
    .page(URL)
    .clientScripts({content: SCRIPT, page: URL})
    .requestHooks(logger)
    ('Fail', async t => {
        await t.expect(Selector('div').withText('Welcome').visible).ok()
        await t.expect(logger.requests.length).eql(1)
    });
```

**Comment by Klaster1** [comment]
@Dmitry-Ostashev wouldn't `test.requestHooks` make independent hook instances for each test way more complex to implement compared to `before`/`after`, where I would usually init and dispose of hooks (free up ports, server instances, etc.) in async manner? Is there a way to reliably init a hook before page has been initialized, besides `t.navigateTo` instead of `test.page`?

If there's no other way, I suggest you to at least consider an extra paragraph on hook/test lifecycle, I read the hook documentation and the closest to what you say can be found [here](https://devexpress.github.io/testcafe/documentation/test-api/intercepting-http-requests/attaching-hooks-to-tests-and-fixtures.html):

> You can use the t.addRequestHooks and t.removeRequestHooks methods to attach and detach hooks _during the test_.

And in test docs [here](https://devexpress.github.io/testcafe/documentation/test-api/test-code-structure.html#test-hooks):

> At the moment test hooks run, the tested webpage is already loaded

Only after you explained the issue the two parts clicked together, this nuance could have been more obvious. I'm am partially to blame for not reading the docs super carefully, but that also points a documentation gap that's easy to address.

Still, it would be nice to have a way to add a hook with access to test context (i.e. `t.addRequestHooks`) before the very first navigation happens, the way I originally expect for it to work.

**Comment by Farfurix** [comment]
@Klaster1 

Hello,

Thank you for your request. We will take it into account.

> wouldn't `test.requestHooks` make independent hook instances for each test way more complex to implement compared to `before`/`after`, where I would usually init and dispose of hooks (free up ports, server instances, etc.) in async manner?

Can you please describe your case and provide test code examples?

> Is there a way to reliably init a hook before page has been initialized, besides `t.navigateTo` instead of `test.page`?

I checked [the above example](https://github.com/DevExpress/testcafe/issues/4616#issuecomment-570532048) and it passes without any assertion errors. Is it possible to use this approach in your project?

**Comment by Klaster1** [comment]
@Farfurix

> Can you please describe your case and provide test code examples?

You can glipse of what real project test code looks like [here](https://github.com/gridgain/gridgain/blob/wc-master/modules/web-console/e2e/testcafe/fixtures/queries/notebook.js). Keep in mind that this is only a small public portion of code. My team wants to use hooks and concurrent, isolated tests to improve test suite run time, but issues like this introduce unnecessary complications.

> I checked the above example and it passes without any assertion errors. Is it possible to use this approach in your project?

The `navigateTo` workaround I mentioned above solves the issue, which in turn means current TestCafe does not necessary require any additional features/fixes. I think you can still action on this, either by improving documentation so other developers don't encounter the same issue I did, and/or change/improve how hooks and `test.page` mesh together to enable more streamlined approach to cases similar to mine (i.e. need hooks reliably ready before test execution start or inital page navigation).

**Comment by AndreyBelym** [comment]
As I understand, your original inquiry can be resolved by using `test.requestHooks` and `fixture.requestHooks`. So I close this issue.

While you can use the [userAgent](https://devexpress.github.io/testcafe/documentation/test-api/intercepting-http-requests/logging-http-requests.html#the-request-object) property to figure out which browser generated a specific request log and use it as a filter criterion, I understand that it can be not an obvious and convenient way to handle this task. So I created a separate issue to discuss and track our progress in implementing your suggestion: https://github.com/DevExpress/testcafe/issues/4624.
